File: examples/android-map/src/main/kotlin/org/maplibre/nativeffi/examples/androidmap/Channels.kt
package org.maplibre.nativeffi.examples.androidmap

import java.util.concurrent.atomic.AtomicBoolean
import org.maplibre.nativeffi.geo.ScreenPoint

/**
 * A camera change decoded on the render loop and applied on the map's owner thread.
 *
 * Commands carry deltas rather than absolute targets, because reading the camera and writing the
 * new one has to happen together on the thread that owns the map. Coordinates are logical map
 * units, converted by the render loop that owns the viewport.
 */
internal sealed interface CameraCommand {
  data object CancelTransitions : CameraCommand

  data class SetGestureInProgress(val inProgress: Boolean) : CameraCommand

  data class MoveBy(val deltaX: Double, val deltaY: Double) : CameraCommand

  data class ScaleBy(val scale: Double, val anchor: ScreenPoint) : CameraCommand

  data class AdjustBearing(val degrees: Double, val anchor: ScreenPoint) : CameraCommand

  data class AdjustPitch(val degrees: Double) : CameraCommand

  /** Double tap: animate to one zoom level past the nearest whole level, about [anchor]. */
  data class ZoomToNextWholeLevel(val anchor: ScreenPoint) : CameraCommand
}

/**
 * One-bit signal that a frame is worth drawing. The render loop consumes before it renders and sets
 * again when nothing was rendered, so a request published during a render is not lost.
 */
internal class RenderRequest {
  private val value = AtomicBoolean(true)

  fun set() {
    value.set(true)
  }

  fun consume(): Boolean = value.getAndSet(false)
}

File: examples/dotnet-map/Channels.cs
using Maplibre.NativeFfi.Camera;
using Maplibre.NativeFfi.Geo;

namespace Maplibre.NativeFfi.Examples.DotnetMap;

internal abstract record CameraCommand;

internal sealed record CancelTransitionsCommand : CameraCommand;

internal sealed record SetGestureInProgressCommand(bool InProgress) : CameraCommand;

internal sealed record MoveByCommand(double DeltaX, double DeltaY, AnimationOptions? Animation)
    : CameraCommand;

internal sealed record ScaleByCommand(
    double Scale,
    ScreenPoint? Anchor,
    AnimationOptions? Animation
) : CameraCommand;

internal sealed record AdjustBearingCommand(double Delta, AnimationOptions? Animation)
    : CameraCommand;

internal sealed record AdjustPitchCommand(double Delta, AnimationOptions? Animation)
    : CameraCommand;

internal sealed record ResetOrientationCommand(AnimationOptions Animation) : CameraCommand;

/// <summary>Submits decoded input directly through the map's any-thread command API.</summary>
internal sealed class CommandQueue(MapState state)
{
    public void Push(CameraCommand command)
    {
        _ = state.Apply(command);
    }
}

/// <summary>One-bit signal that a frame is worth drawing.</summary>
internal sealed class RenderRequest
{
    private int value = 1;

    public void Set()
    {
        Volatile.Write(ref value, 1);
    }

    public bool Consume()
    {
        return Interlocked.Exchange(ref value, 0) == 1;
    }
}
